libsolidity/analysis/ControlFlowGraph.cpp

```diff
@@ -0,0 +1,552 @@
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <libsolidity/analysis/ControlFlowGraph.h>
+
+#include <boost/range/adaptor/reversed.hpp>
+
+#include <algorithm>
+
+using namespace std;
+using namespace dev::solidity;
+
+namespace dev
+{
+namespace solidity
+{
+
+/** Helper class that parses the control flow of a non-top-level ASTNode. */
+class ControlFlowParser: private ASTConstVisitor
+{
+public:
+	/// Connects @a _entry and @a _exit by inserting the control flow of @a _node in between.
+	static void createFlowFromTo(CFG& _cfg, CFGNode* _entry, CFGNode* _exit, ASTNode const& _node)
+	{
+		ControlFlowParser parser(_cfg, _entry);
+		parser.appendControlFlow(_node);
+		connect(parser.m_currentNode, _exit);
+	}
+
+private:
+	explicit ControlFlowParser(CFG& _cfg, CFGNode* _entry):
+		m_cfg(_cfg), m_currentNode(_entry) {}
+
+	virtual bool visit(BinaryOperation const& _operation) override;
+	virtual bool visit(Conditional const& _conditional) override;
+	virtual bool visit(IfStatement const& _ifStatement) override;
+	virtual bool visit(ForStatement const& _forStatement) override;
+	virtual bool visit(WhileStatement const& _whileStatement) override;
+	virtual bool visit(Break const&) override;
+	virtual bool visit(Continue const&) override;
+	virtual bool visit(Throw const&) override;
+	virtual bool visit(Block const&) override;
+	virtual void endVisit(Block const&) override;
+	virtual bool visit(Return const& _return) override;
+	virtual bool visit(PlaceholderStatement const&) override;
+	virtual bool visit(FunctionCall const& _functionCall) override;
+
+
+	/// Appends the control flow of @a _node to the current control flow.
+	void appendControlFlow(ASTNode const& _node)
+	{
+		_node.accept(*this);
+	}
+
+	/// Starts at @a _entry and parses the control flow of @a _node.
+	/// @returns The node at which the parsed control flow ends.
+	/// m_currentNode is not affected (it is saved and restored).
+	CFGNode* createFlow(CFGNode* _entry, ASTNode const& _node)
+	{
+		auto oldCurrentNode = m_currentNode;
+		m_currentNode = _entry;
+		appendControlFlow(_node);
+		auto endNode = m_currentNode;
+		m_currentNode = oldCurrentNode;
+		return endNode;
+	}
+
+	/// Creates an arc from @a _from to @a _to.
+	static void connect(CFGNode* _from, CFGNode* _to)
+	{
+		solAssert(_from, "");
+		solAssert(_to, "");
+		_from->exits.push_back(_to);
+		_to->entries.push_back(_from);
+	}
+
+
+protected:
+	virtual bool visitNode(ASTNode const& node) override;
+
+private:
+
+	/// Splits the control flow starting at the current node into n paths.
+	/// m_currentNode is set to nullptr and has to be set manually or
+	/// using mergeFlow later.
+	template<size_t n>
+	std::array<CFGNode*, n> splitFlow()
```

Comment:
Does this need to be a template?

Reply:
Ok, it does not hurt because it is a cpp file anyway.

Reply:
Yeah. The number of paths to split into is always known at compile time and this function is only used locally in this cpp file, so I would say a template makes sense. However, it turns out that so far it's actually only used with ``n = 2`` in the end, so I could also just fix that and make it return a ``pair`` instead. I thought it made sense to keep it general, though.

Reply:
It's fine!